internal/cluster/store/store.go

```diff
@@ -0,0 +1,855 @@
+package store
+
+import (
+	"errors"
+	"expvar"
+	"fmt"
+	"net"
+	"os"
+	"path/filepath"
+	"sort"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/raft"
+	raftboltdb "github.com/hashicorp/raft-boltdb"
+	"github.com/pbinitiative/zenbpm/internal/cluster/command/proto"
+	zproto "github.com/pbinitiative/zenbpm/internal/cluster/proto"
+	"github.com/pbinitiative/zenbpm/internal/config"
+	"github.com/rqlite/rqlite/v8/random"
+	"github.com/rqlite/rqlite/v8/rsync"
+	"github.com/rqlite/rqlite/v8/tcp"
+	pb "google.golang.org/protobuf/proto"
+)
+
+const (
+	observerChanLen     = 100
+	connectionPoolCount = 5
+	connectionTimeout   = 10 * time.Second
+	leaderWaitDelay     = 100 * time.Millisecond
+)
+
+const (
+	numSnapshots              = "num_snapshots"
+	numSnapshotsFailed        = "num_snapshots_failed"
+	numBackups                = "num_backups"
+	numRestores               = "num_restores"
+	numRestoresFailed         = "num_restores_failed"
+	numJoins                  = "num_joins"
+	numIgnoredJoins           = "num_ignored_joins"
+	numRemovedBeforeJoins     = "num_removed_before_joins"
+	snapshotCreateDuration    = "snapshot_create_duration"
+	numSnapshotPersists       = "num_snapshot_persists"
+	numSnapshotPersistsFailed = "num_snapshot_persists_failed"
+	snapshotPersistDuration   = "snapshot_persist_duration"
+	leaderChangesObserved     = "leader_changes_observed"
+	leaderChangesDropped      = "leader_changes_dropped"
+	failedHeartbeatObserved   = "failed_heartbeat_observed"
+	nodesReapedOK             = "nodes_reaped_ok"
+	nodesReapedFailed         = "nodes_reaped_failed"
+)
+
+var (
+	// stats captures stats for the Store.
+	stats *expvar.Map
+)
+
+func init() {
+	stats = expvar.NewMap("zenbpm-store")
+	ResetStats()
+}
+
+func ResetStats() {
+	stats.Init()
+	stats.Add(numSnapshots, 0)
+	stats.Add(numSnapshotsFailed, 0)
+	stats.Add(numBackups, 0)
+	stats.Add(numRestores, 0)
+	stats.Add(numRestoresFailed, 0)
+	stats.Add(numJoins, 0)
+	stats.Add(numIgnoredJoins, 0)
+	stats.Add(numRemovedBeforeJoins, 0)
+	stats.Add(snapshotCreateDuration, 0)
+	stats.Add(numSnapshotPersists, 0)
+	stats.Add(numSnapshotPersistsFailed, 0)
+	stats.Add(snapshotPersistDuration, 0)
+	stats.Add(leaderChangesObserved, 0)
+	stats.Add(leaderChangesDropped, 0)
+	stats.Add(failedHeartbeatObserved, 0)
+	stats.Add(nodesReapedOK, 0)
+	stats.Add(nodesReapedFailed, 0)
+}
+
+type Store struct {
+	config Config
+
+	open *atomic.Bool
+
+	notifyMu        sync.Mutex
+	bootstrapExpect int
+	bootstrapped    bool
+	notifyingNodes  map[string]raft.Server
+
+	mu     sync.Mutex
```

Comment:
What scope does this mutex have?
I just see usage in the fsm.go and assume the mutex should guard access to the state field, but not all the other fields in this struct.
IF so, I recommend renaming to `stateMutex` or similar, to indicate the scope (meaning when shall it be used.